### PR TITLE
Improve and fix FFI tests

### PR DESCRIPTION
**Pull request by ollydev**
- Previously FFI tests exit code was ignored. Now if a test fails the Travis job fails too.
- Added static linking build modes and tests (`default-static` `cdecl-static` etc)
- Added OSX static link binary 
- Added Linux static link binary
- In tests that use `Natify` use `TEST_ABI` . Previously it was using the default `FFI_DEFAULT_ABI`
- Fix `Extended` datatype tests when static linking on `win64`
- Add tests for power of 2 size records (2 to 64).
- Fixed bugs related to smaller size records. I can't say I did any research into the calling conventions but the issues and fixes were similar to previous patches.

And now.... ALL THE TESTS PASS!

### CODE REVIEW

File: extensions/ffi/lpffiwrappers.pas [hunk 1]
```diff
@@ -822,7 +827,7 @@ function LapeFFIPointerParam(const Param: TLapeParameter; ABI: TFFIABI): Boolean
 
   {$IF DECLARED(FFI_WIN64)}
       or ((ABI = FFI_WIN64) and (Param.VarType.BaseType = ltRecord) and (not (UInt8(Param.VarType.Size) in PowerTwoRegs)))
-      or ((ABI = FFI_WIN64) and (Param.VarType.BaseType = ltRecord) and (Param.ParType in Lape_ConstParams) and (Param.VarType.Size = SizeOf(UInt8)))
+      or ((ABI = FFI_WIN64) and (Param.VarType.BaseType = ltRecord) and (Param.ParType in Lape_ConstParams) and (Param.VarType.Size in PowerTwoRegs))
```
Comment (nielsAD) [inline extensions/ffi/lpffiwrappers.pas]: You can add `or (Param.ParType in Lape_ConstParams)` to the previous statement, as they almost check for the same condition.

File: extensions/ffi/lpffiwrappers.pas [hunk 2]
```diff
@@ -833,7 +838,7 @@ function LapeFFIStackParam(const Param: TLapeParameter; ABI: TFFIABI): Boolean;
     Result := (Param.VarType <> nil) and (Param.VarType.BaseType = ltDynArray)
 
     {$IF (FPC_VERSION >= 3) AND DECLARED(FFI_PASCAL)}
-      or ((ABI = FFI_PASCAL) and (Param.VarType.Size < SizeOf(Pointer)) and (Param.VarType.Size <> SizeOf(UInt16)))
+      or ((ABI = FFI_PASCAL) and (Param.VarType.Size < SizeOf(Pointer)));
```
Comment (nielsAD) [inline extensions/ffi/lpffiwrappers.pas]: Semicolon not needed here, there's one below. Also not sure why the UInt16 size check was here before, but I'll trust  the tests.

**Comment by ollydev** [comment]
@nielsAD Done!

**Comment by nielsAD** [comment]
Thanks!

**Pull request merged**